src/librustc/ty/mod.rs

```diff
@@ -1034,7 +1034,7 @@ impl<'tcx> ToPolyTraitRef<'tcx> for PolyProjectionPredicate<'tcx> {
         // This is because here `self` has a `Binder` and so does our
         // return value, so we are preserving the number of binding
         // levels.
-        ty::Binder(self.0.projection_ty.trait_ref)
+        ty::Binder(ty::tls::with(|tcx| self.0.projection_ty.trait_ref(tcx)))
```

Comment:
No worries, I track all comments that I haven't answered in the affirmative.

Removing the impl breaks, among other places, `process_predicate`:

```rust
        ty::Predicate::Projection(ref data) => {
            let project_obligation = obligation.with(data.clone());
            match project::poly_project_and_unify_type(selcx, &project_obligation) {
                Ok(None) => {
                    pending_obligation.stalled_on =
                        trait_ref_type_vars(selcx, data.to_poly_trait_ref());
                    Ok(None)
                }
                Ok(v) => Ok(v),
                Err(e) => Err(CodeProjectionError(e))
            }
        }
```